Port Shujing's I-Surprisingness to URE pattern miner

ngeiswei opened the pull request.

ngeiswei committed:
Simplify comb calculation by using boost::math::binomial_coefficient

ngeiswei committed:
Add get/set count to pattern

ngeiswei committed:
More newlines

ngeiswei committed:
Pass texts by reference instead of copy in Miner::specialize_shapat

ngeiswei committed:
Store support in the pattern

ngeiswei committed:
Initial implementation of I-Surprisingness

ngeiswei committed:
Misc fixes

ngeiswei committed:
Fix I-Surprisingness

ngeiswei committed:
Unary shallow specialization when conjunction expansion is enabled

To not circumvent the heuristic of conjunction expansion

ngeiswei committed:
Disable I-Surprisingness testing till fixed

ngeiswei committed:
Simplify miner-utils by implementing iota-one, [1, n] instead of [0, n)

ngeiswei committed:
Minor code improvements

ngeiswei committed:
Move most Miner UTest helpers to MinerUTestUtils

ngeiswei committed:
Some minor function renaming

ngeiswei committed:
Load miner-rule-utils in I-Surprisingness rule

ngeiswei committed:
Misc improvements in MinerUTestUtils

ngeiswei committed:
Add SurprisingnessUTest

ngeiswei committed:
Use new pattern matcher API

ngeiswei committed:
Fix MinerUTestUtils::add_soda_drinker_pattern

ngeiswei committed:
Restore SurprisingnessUTest after fixing soda drinker helper

ngeiswei committed:
Use binonial instead power for backward compatibility with Shunjing's

ngeiswei committed:
Test surprisingness evaluation is absent to avoid redundant calc

ngeiswei committed:
Support partitions in I-Surprisingness

ngeiswei committed:
Add partitions documentation

ngeiswei committed:
Only update TV is different than default

ngeiswei committed:
Enable I-Surprisingness utest

ngeiswei committed:
Add note about bug reproduction

vsbogd commented on opencog/learning/miner/MinerUtils.h:
```diff
@@ -421,6 +421,24 @@ class MinerUtils
+	/**
+	 * Return an atom to serve as key to store the support value.
+	 */
+	static const Handle& support_key();
```
Minor, but may be hide this method and move it to private part of the interface.

ngeiswei replied:
Good point, though I'm not sure it should be moved to a private part because I might add a scheme binding to it in the future, to grab pattern support, etc. Alternatively I could add a binding for `get_pattern`, etc, I'm just not sure which road I'm gonna take yet.

vsbogd commented on opencog/learning/miner/CMakeLists.txt:
```diff
@@ -3,6 +3,7 @@ ADD_LIBRARY(miner SHARED
 	MinerUtils
 	HandleTree
 	Valuations
+   Surprisingness
```
```suggestion
	Surprisingness
```

vsbogd commented on opencog/learning/miner/Surprisingness.cc:
```diff
@@ -0,0 +1,116 @@
+/*
+ * Miner.cc
```
```suggestion
 * Surprisingness.cc
```

vsbogd commented on opencog/learning/miner/Surprisingness.cc:
```diff
@@ -0,0 +1,116 @@
+/*
+ * Miner.cc
+ *
+ * Copyright (C) 2017 OpenCog Foundation
```
```suggestion
 * Copyright (C) 2019 SingularityNET Foundation
```

ngeiswei committed:
Rename I-Surprisingness predicate by isurp + remove conjuncts clause

The conjuncts clauses seems to not be useful as their count can always
be recomputed if missing.

ngeiswei commented:
Thanks for the feedback @vsbogd , I'm closing for now as I realized it needs more work anyway.
